cli/src/commands/gerrit/upload.rs

```diff
@@ -0,0 +1,389 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+use std::io::Write as _;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use bstr::BStr;
+use indexmap::IndexMap;
+use itertools::Itertools as _;
+use jj_lib::backend::CommitId;
+use jj_lib::commit::Commit;
+use jj_lib::commit::CommitIteratorExt as _;
+use jj_lib::git::GitRefUpdate;
+use jj_lib::git::{self};
+use jj_lib::object_id::ObjectId as _;
+use jj_lib::repo::Repo as _;
+use jj_lib::revset::RevsetExpression;
+use jj_lib::settings::UserSettings;
+use jj_lib::store::Store;
+use jj_lib::trailer::Trailer;
+use jj_lib::trailer::parse_description_trailers;
+
+use crate::cli_util::CommandHelper;
+use crate::cli_util::RevisionArg;
+use crate::cli_util::short_commit_hash;
+use crate::command_error::CommandError;
+use crate::command_error::internal_error;
+use crate::command_error::user_error;
+use crate::command_error::user_error_with_hint;
+use crate::command_error::user_error_with_message;
+use crate::git_util::with_remote_git_callbacks;
+use crate::ui::Ui;
+
+#[derive(clap::Args, Clone, Debug)]
+pub struct UploadArgs {
+    /// The revset, selecting which revisions are sent in to Gerrit. This can be
+    /// any arbitrary set of commits.
+    /// Note that when you push a commit at the head of a stack, all ancestors
+    /// are pushed too. This means that `jj gerrit upload -r foo` is equivalent
+    /// to `jj gerrit upload -r 'mutable()::foo`
+    #[arg(long, short = 'r')]
+    revisions: Vec<RevisionArg>,
+
+    /// The location where your changes are intended to land. This should be
+    /// an upstream branch.
+    #[arg(long = "remote-branch", short = 'b')]
+    remote_branch: Option<String>,
+
+    /// The Gerrit remote to push to. Can be configured with the `gerrit.remote`
+    /// repository option as well. This is typically a full SSH URL for your
+    /// Gerrit instance.
+    #[arg(long)]
+    remote: Option<String>,
+
+    /// If true, do not actually push the changes to Gerrit.
+    #[arg(long = "dry-run", short = 'n')]
+    dry_run: bool,
+}
+
+/// calculate push remote. The logic is:
+/// 1. If the user specifies `--remote`, use that
+/// 2. If the user has 'gerrit.remote' configured, use that
+/// 3. If there is a default push remote, use that
+/// 4. If the user has a remote named 'gerrit', use that
+/// 5. otherwise, bail out
```

Comment:
nit: This sounds like something that should be explained in user-facing docs. (Or, if it's clear enough that users don't need that documentation, then I don't think we need it here either.)